### PR TITLE
docs: aa-core withGasEstimator

### PR DESCRIPTION
**Pull request by avarobinson**
docs: add aa-core withGasEstimator

### CODE REVIEW

File: site/packages/aa-core/provider/withGasEstimator.md [hunk 1]
```diff
@@ -0,0 +1,47 @@
+---
+outline: deep
+head:
+  - - meta
+    - property: og:title
+      content: withGasEstimator
```
Comment (denniswon) [inline site/packages/aa-core/provider/withGasEstimator.md:6]:
```suggestion
      content: ISmartAccountProvider • withGasEstimator
```

File: site/packages/aa-core/provider/withGasEstimator.md [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+    - name: description
+      content: Overview of the withGasEstimator method on ISmartAccountProvider
+  - - meta
+    - property: og:description
+      content: Overview of the withGasEstimator method on ISmartAccountProvider
+---
+
+# withGasEstimator
+
+Override the gasEstimator middleware which is used for setting the gasLimit fields on the UserOperation prior to execution.
```
Comment (denniswon) [inline site/packages/aa-core/provider/withGasEstimator.md:17]: Include reference links. Better with more explanation how gas estimation works.

File: site/packages/aa-core/provider/withGasEstimator.md [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+
+```ts [example.ts]
+import { provider } from "./provider";
+
+// [!code focus:99]
+provider.withGasEstimator(async () => ({
+  callGasLimit: 0n,
+  preVerificationGas: 0n,
+  verificationGasLimit: 0n,
+}));
```
Comment (denniswon) [inline site/packages/aa-core/provider/withGasEstimator.md:31]: Can we have more practical example? Don't think people will be able to figure out how to set these override funcs

**Review by denniswon — CHANGES_REQUESTED** [review]

**Comment by SahilAujla** [comment]
We will also need to mention `gasEstimator` in the same page as that's the related readOnly field and we decided to mention them along with these here.

For example, see `feeDataGetter` for `withFeeDataGetter`: https://github.com/SahilAujla/aa-sdk-private/blob/withFeeDataGetter/site/packages/aa-core/provider/withFeeDataGetter.md#feedatagetter

File: site/packages/aa-core/provider/withGasEstimator.md (hunk 3 above)
Comment (avasisht23) [reply site/packages/aa-core/provider/withGasEstimator.md:31]: +1

For instance, after setting the middleware, maybe call buildUserOperation and show in a comment how resulting uo struct fields (callGasLimit, pvg, vgl) changed in the resulting uo struct after build due to this middleware override

File: site/packages/aa-core/provider/withGasEstimator.md [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+```
+
+<<< @/snippets/provider.ts
+:::
+
+## Returns
+
+### `SmartAccountProvider`
+
+The provider with the gasEstimator middleware override.
```
Comment (avasisht23) [inline site/packages/aa-core/provider/withGasEstimator.md:41]:
```suggestion
The provider with the `gasEstimator` middleware override.
```

**Review by avasisht23 — COMMENTED** [review]
can you make the example more detailed? thanks!

File: site/packages/aa-core/provider/withGasEstimator.md (hunk 3 above)
Comment (avasisht23) [reply site/packages/aa-core/provider/withGasEstimator.md:31]: this wasn't addressed, but was looking for:

```suggestion
provider.withGasEstimator(async () => ({
  callGasLimit: 0n,
  preVerificationGas: 0n,
  verificationGasLimit: 0n,
}));

// Define the user operation data
const userOpData = {
  target: "0xTARGET_ADDRESS", // Replace with your actual target address
  data: "0xSOME_DATA", // Replace with your actual data
};

const resultingUO = await provider.buildUserOperation(userOpData);
console.log("Modified callGasLimit:", resultingUO.callGasLimit);
console.log("Modified preVerificationGas:", resultingUO.verificationGasLimit);
console.log("Modified verificationGasLimit:", resultingUO.verificationGasLimit);
```

**Review by avasisht23 — APPROVED** [review]
can you add the example build UO to the usage section? then can ship

**Comment by avarobinson** [comment]
> can you add the example build UO to the usage section? then can ship

@avasisht23 just updated the example and read only fields

**Review by moldy530 — APPROVED** [review]

**Pull request merged**